### PR TITLE
Installer script

### PR DESCRIPTION
**Pull request by chad-ermacora**
Interactive install script for darkstar on Debian, Ubuntu & Raspbian.  Only need to enter a username, password and IP to get up and running.  The user is used to start the servers and access the SQL database.

### CODE REVIEW

File: darkstar_install_script.sh [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+  # shellcheck disable=SC2086
+  apt-get update && apt-get -y install ${APT_GET_INSTALL_ONE} ${APT_GET_INSTALL_TWO}
+  adduser "${DARKSTAR_USER}" --gecos "Darkstar Server" --disabled-password
+  if [[ ${GIT_VERSION} == "1" ]]; then
+    git clone --branch stable --recursive https://github.com/DarkstarProject/darkstar/ /opt/darkstar
+  else
+    git clone --recursive https://github.com/DarkstarProject/darkstar/ /opt/darkstar
+  fi
+  cd /opt/darkstar || exit
+  sh autogen.sh
```
Comment (whasf) [inline darkstar_install_script.sh]: No check for if this is successful or not?

File: darkstar_install_script.sh [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  if [ "${OS}" == "debian" ] || [ "${OS}" == "raspbian" ]; then
+    printf "\nRunning Debian or Raspbian Install\n"
+    apt-get -y install libmariadb-dev-compat
+    ./configure CXXFLAGS=" -pthread"
```
Comment (whasf) [inline darkstar_install_script.sh]: Check if it completes successfully?

File: darkstar_install_script.sh [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  else
+    printf "\nRunning Default Install\n"
+    ./configure
```
Comment (whasf) [inline darkstar_install_script.sh]: Check if it completes successfully?

File: darkstar_install_script.sh [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  fi
+  make -j4
```
Comment (whasf) [inline darkstar_install_script.sh]: Check if it completes successfully?

File: darkstar_install_script.sh [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+  chown -R "${DARKSTAR_USER}":root /opt/darkstar/
+  printf "\nEnter the MySQL root password\nDefault: No password\n"
+  mysql -h "localhost" -u "root" -p -e "CREATE USER '${DARKSTAR_USER}'@'localhost' IDENTIFIED BY '${DARKSTAR_PASSWORD}';CREATE DATABASE dspdb;USE dspdb;GRANT ALL PRIVILEGES ON dspdb.* TO '${DARKSTAR_USER}'@'localhost';"
```
Comment (whasf) [inline darkstar_install_script.sh]: Is the service running? (Does the install automatically start it?)

File: darkstar_install_script.sh (hunk 2 above)
Comment (chad-ermacora) [reply darkstar_install_script.sh]: for the apt-get install or the ./configure?

File: darkstar_install_script.sh (hunk 5 above)
Comment (chad-ermacora) [reply darkstar_install_script.sh]: If you mean the 3 darkstar services (connect, game, search)they are automatically enabled & started at the end of the script.  The SQL database is also up and running right after it's installed with apt-get above.

File: darkstar_install_script.sh (hunk 3 above)
Comment (chad-ermacora) [reply darkstar_install_script.sh]: Do the ./configure, make and autogen.sh return a none 0 value when they fail? I should be able to add a `|| exit` if they do.

File: darkstar_install_script.sh (hunk 1 above)
Comment (chad-ermacora) [reply darkstar_install_script.sh]: Now exits on failure.

File: darkstar_install_script.sh (hunk 2 above)
Comment (chad-ermacora) [reply darkstar_install_script.sh]: Now exits on failure.

File: darkstar_install_script.sh (hunk 3 above)
Comment (chad-ermacora) [reply darkstar_install_script.sh]: I was able to test it, and it does indeed exit on fail with `|| exit` and continues on success. 
Now set to exit on failure.

File: darkstar_install_script.sh (hunk 4 above)
Comment (chad-ermacora) [reply darkstar_install_script.sh]: Now exits on failure.

File: darkstar_install_script.sh (hunk 5 above)
Comment (chad-ermacora) [reply darkstar_install_script.sh]: I have clarified the descriptive text for services, explaining how it set's the server to auto-start or disables auto-start.

File: darkstar_install_script.sh [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+  echo
+  if [[ ${ZONEIP} =~ ^[Yy]$ ]]; then
+    printf "\nYour current systems IPv4 Address are as follows\n"
+    ip -4 addr | grep -oP '(?<=inet\s)\d+(\.\d+){3}'
+    echo
+    read -r -p "Enter the IP address you wish to use with this Server: " DARKSTAR_IP
+    printf "\nEnter the MySQL root password\nDefault: No password\n"
+    mysql -h "localhost" -u "root" -p -e "USE dspdb;UPDATE zone_settings SET zoneip = '${DARKSTAR_IP}';"
+  fi
+  nano ${INSTALL_DIR}/conf/login_darkstar.conf
```
Comment (ShelbyZ) [inline darkstar_install_script.sh]: Is the purpose of opening these files to allow the users to make changes?

**Comment by ShelbyZ** [comment]
There are some potential problems with what this script will offer:

- We are not providing the ability for a user that already has this installed to update branches or code via a 'friendly' option
- How do we plan on handling dependency changes
- How do we plan on handling SQL changes

Feels like there should be an option for updating the server from GIT and figuring out how those changes would be applied.  Code would be easy, SQL would require usage of migration scripts.

File: darkstar_install_script.sh (hunk 6 above)
Comment (chad-ermacora) [reply darkstar_install_script.sh]: That is correct.  You have to set the newly created user and password for the SQL database access.

**Comment by chad-ermacora** [comment]
> There are some potential problems with what this script will offer:
> 
> * We are not providing the ability for a user that already has this installed to update branches or code via a 'friendly' option
> * How do we plan on handling dependency changes
> * How do we plan on handling SQL changes
> 
> Feels like there should be an option for updating the server from GIT and figuring out how those changes would be applied. Code would be easy, SQL would require usage of migration scripts.

I was told to not allow upgrades, due to SQL change issues that are currently not automated. So if install is chosen but it detects it has been built already, it exits the script.  Configurations, such as Zone IP and the 3 main configuration files can still be accessed after setup by running the script and choosing edit config option. 

Dependency changes will have to be updated in the script by adding or removing it in one of the 2 APT_GET_INSTALL variables near the top. If they are Distro specific, they can be added in the section shown below. Adding other distro's should be relatively easy, as you add in an extra else if statement checking for the lowercase name of the distro.  

`if [ "${OS}" == "debian" ] || [ "${OS}" == "raspbian" ]; then
    printf "\nRunning Debian or Raspbian Install\n"
    apt-get -y install libmariadb-dev-compat
    ./configure CXXFLAGS=" -pthread" || exit
  else
    printf "\nRunning Default Install\n"
    ./configure || exit
  fi`

If we did have a SQL script in the main repository that would update SQL entries based on whatever has changed, I could link directly to it in the script and download it at runtime.  I was thinking as long as the SQL script is updated before pushing to Stable, upgrading between stable releases should be a viable option.

**Comment by TeoTwawki** [comment]
as long as they can git pull after they can get code changes and then rebuild for themselves. 

we can't really get a good way to automate sql changes at this time.

File: darkstar_install_script.sh (hunk 6 above)
Comment (chad-ermacora) [reply darkstar_install_script.sh]: I have now added text to explain why said files will be opened and how to use the text editor.

**Comment by chad-ermacora** [comment]
-Added fix to work on Ubuntu desktop (I was testing on server, which enables required repository by default but desktop does not)
-Misc text output improvements.

**Pull request merged**